Switched type to `unique` to be consistent with Mongoose types

niftylettuce opened the pull request:
Closes #39

niftylettuce commented:
Fixing tests now

niftylettuce commented:
@matteodelabre I have fixed the tests and force-pushed after rebasing, should be good now.

matteodelabre commented:
Hello @niftylettuce, and thanks for your contribution! I’m currently reviewing this.

matteodelabre commented on index.js:
```diff
@@ -72,7 +72,8 @@ function beautify(error, collection, values, messages) {
                     var props = {
-                        type: 'Duplicate value',
+                        type: 'unique',
+                        kind: 'unique',
```
Looks like `Mongoose.ValidatorError` only considers the `type` property from the properties object and [internally converts it to a `kind` attribute](https://github.com/Automattic/mongoose/blob/4.11.1/lib/error/validator.js#L30). So I don’t think we need to supply a `kind` attribute.

niftylettuce replied:
@matteodelabre The tests wouldn't pass without having `kind` 😦

matteodelabre replied:
I think this is why: https://github.com/matteodelabre/mongoose-beautiful-unique-validation/pull/40#discussion_r131784417 — still, we shouldn’t keep the duplicated attribute IMHO.

niftylettuce replied:
Will push and send you a message shortly @matteodelabre doing some tests with project using it atm, I added something else to add a message formatter to the string, will share in a bit

matteodelabre commented on index.js:
```diff
@@ -72,13 +72,14 @@ function beautify(error, collection, values, messages) {
 
                     if (typeof messages[path] === 'string') {
-                        props.message = messages[path];
+                        props.message = props.reason = messages[path];
```
Looks like the `reason` attribute [is not being used by most validators](https://github.com/Automattic/mongoose/blob/4.11.1/lib/schema/string.js#L108), so do you think we should provide it? If so, I’m not sure whether it’s appropriate to set it to the same value as `message`.

niftylettuce replied:
I can get rid of it - just let me know about my other comment above <https://github.com/matteodelabre/mongoose-beautiful-unique-validation/pull/40#discussion_r131781824> and I'll do it all at once.

matteodelabre commented on tests/save.js:
```diff
@@ -11,14 +11,15 @@ mongoose.plugin(beautifulValidation);
-function assertUniqueError(assert, err, dups, message) {
-    assert.equal(err.name, 'ValidationError',
+function assertUniqueError(t, err, dups, message) {
+
```
Looks like you left a blank line here. Just a nitpick :)

niftylettuce replied:
sorry will fix

matteodelabre commented on tests/save.js:
```diff
@@ -11,14 +11,15 @@ mongoose.plugin(beautifulValidation);
  * Assert that the given object is a correct validation error.
  *
- * @param {Object} assert Tape assertion object
+ * @param {Object} t Tape assertion object
```
Actually, renaming the assertion object is quite a good idea. It makes the tests more terse. Thanks!

niftylettuce replied:
thx

matteodelabre commented on tests/save.js:
```diff
@@ -28,47 +29,48 @@ function assertUniqueError(assert, err, dups, message) {
-            'the sub-error\'s type should be "Duplicate value"'
+        t.equal(
+            suberr.properties.kind, 'unique',
+            'the sub-error\'s kind should be "unique"'
```
This is causing tests failure when you remove the `type` attribute from the `properties` object, because `suberr`’s own attributes are those translated by the ValidatorError constructor, and `suberr.properties` is a reference to the object passed in argument to that constructor.

I don’t actually understand `suberr.properties` purpose (and it seems redundant with `suberr`’s own attributes). We are left with two choices: removing this assertion altogether and ignoring `suberr.properties`, or using `suberr.properties.type` instead of `suberr.properties.kind` in this assertion.

The second choice looks best to me.

niftylettuce commented:
@matteodelabre OK I've updated my PR per all your requests - please see my latest push.  Note that I've added a fallback global message formatted adhering to Mongoose's standard behavior (e.g. <https://github.com/Automattic/mongoose/blob/e659958fe80c5d03a7cb89a4e43804c1dd3ddda2/lib/error/messages.js>).

niftylettuce commented:
Also @matteodelabre I documented `mongoose.Error.messages.general.unique` in the Readme too!

matteodelabre commented:
Looks good to me! Thanks for your help :-)
I will publish a new major release with this as soon as possible.

niftylettuce commented:
I can release an NPM version if you `npm owner add niftylettuce` if needed!

matteodelabre commented:
Let’s do that! But first, I wanted to add some tests for the default validation message and fix some minor formatting issues — unfortunately your latest patch does not pass ESLint validation, but it’s entirely my fault, I didn’t integrate ESLint to the CI. Please give me a few minutes :-)

niftylettuce commented:
Oh awesome!

niftylettuce commented:
@matteodelabre see my comment here <https://github.com/Automattic/mongoose/issues/2284#issuecomment-320810641> 😄

niftylettuce commented:
Not sure what's going on here, but I tried to fix this per https://github.com/matteodelabre/mongoose-beautiful-unique-validation/compare/master...niftylettuce:master yet your master branch here on GitHub is still synced to an older version before you accepted the PR 🤦‍♂️ ....

matteodelabre commented:
Thanks for the comment! Much appreciated. :)

matteodelabre commented:
If you want to submit new changes, you need to create a new pull request, because this one was merged and cannot be re-opened.

niftylettuce commented:
I tried to create a new one... this is weird

niftylettuce commented:
See what I mean here @matteodelabre ?  Never seen this ever before:

<img width="1103" alt="screen shot 2017-08-07 at 9 34 11 pm" src="https://user-images.githubusercontent.com/523141/29052453-339c7c28-7bb8-11e7-9f7b-27f6102bd417.png">

matteodelabre commented:
Strange indeed. What happens if you click “Create pull request”?

niftylettuce commented:
Same thing, this is a GitHub bug 🤦‍♂️

niftylettuce commented:
I sent in a support request on GitHub and tweeted them

matteodelabre commented:
Weird. Please keep me updated if you receive a response, I’m also interested. :-)

niftylettuce commented:
All fixed

On Aug 8, 2017 6:59 AM, "Mattéo Delabre ✏️" <notifications@github.com>
wrote:

> Weird. Please keep me updated if you receive a response, I’m also
> interested. :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matteodelabre/mongoose-beautiful-unique-validation/pull/40#issuecomment-320922526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAf7hSQrKoT0soDDgMcL4bb2CoeV0Pmaks5sWD97gaJpZM4OvxlU>
> .
>
